### ISSUE TITLE
Tracking Issue for checksum freshness

### ISSUE THREAD

**Issue by Xaeroxe** [rust-lang/cargo#14136]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/cargo/issues/14136)*

### Summary

Original issue: #6529 
Implementation: https://github.com/rust-lang/cargo/pull/14137 
Rustc implementation: https://github.com/rust-lang/rust/pull/126930
Documentation: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#checksum-freshness

Add the ability to use content checksums rather than filesystem mtimes to determine if a crate should be recompiled.

### Unresolved Issues

* [ ] How should we handle checksums for files processed by build scripts?
  * One option is to [provide an officially sanctioned API for build scripts.](https://github.com/rust-lang/cargo/issues/12432) Then we can use that API to compute and output the correct checksum value when the relevant function is invoked in the build script.
* [ ] Figure out how this will be stabilized (always on, config)
  * [ ] Get benchmarked https://github.com/rust-lang/cargo/issues/14722
* [ ] If this is on by default, migrate Cargo's tests to it (hack on and make pass)

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

```[tasklist]
### Tasks
- [ ] https://github.com/rust-lang/cargo/pull/14137
- [ ] https://github.com/rust-lang/rust/pull/126930
```

**Comment by weihanglo** [comment]
Here is a result of a performance benchmark that mostly tests the rustc part.

* It adds `aws-sdk-ec2@1.80.0` as a dependency and runs `cargo clean && cargo check`.
* `rg 'checksum:blake3' target/debug/deps | wc -l` shows 8313 matches. It means rustc roughly computes blake3 checksum 8313 times.
* Checksum computations on Cargo side are mostly ignored because Cargo assumes registry are immutable and won't do either mtime or checksum check .

**Result:** On par if you have zero local code.

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `cargo +nightly check --frozen` | 52.986 ± 0.366 | 52.549 | 53.681 | 1.00 |
| `cargo +nightly check --frozen -Zchecksum-freshness` | 53.020 ± 0.315 | 52.442 | 53.480 | 1.00 ± 0.01 |

<details><summary>Benchmark details</summary>
<p>

Setup package

```shell
cargo new foo --lib
cd foo
cargo add aws-sdk-ec2
```

Benchmark script

```shell
hyperfine --warmup 1 \
  --setup "cargo fetch" \
  --min-runs 20 \
  --prepare "cargo clean" \
  --export-markdown result.md \
  "cargo +nightly check --frozen" \
  "cargo +nightly check --frozen -Zchecksum-freshness"
```

Platform information

```
cargo 1.82.0 (8f40fc59f 2024-08-21)
release: 1.82.0
commit-hash: 8f40fc59fb0c8df91c97405785197f3c630304ea
commit-date: 2024-08-21
host: x86_64-unknown-linux-gnu
libgit2: 1.8.1 (sys:0.19.0 vendored)
libcurl: 8.9.0-DEV (sys:0.4.74+curl-8.9.0 vendored ssl:OpenSSL/1.1.1w)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Amazon Linux AMI 2.0.0 [64-bit] (r7a.24xlarge)
```

</p>
</details>

**Comment by weihanglo** [comment]
Here is another performance benchmark that focuses on cargo part.

* Clone https://github.com/awslabs/aws-sdk-rust/commit/505dab66bf0801ca743212678d47d6490d2beba9 and `cargo check -p aws-sdk-ec2` (with pre-filled cache)
* `rg 'checksum:blake3' target/debug/deps | wc -l` shows 8312 matches. So basiclly matches the previous rustc-only benchmark result https://github.com/rust-lang/cargo/issues/14136#issuecomment-2430876726
* Different from the rustc-only benchmark, the entire aws-sdk-ec2 is a local package, so every rust file will be blake-checksum'd by Cargo to check its freshness.

**Result:** performance deteriorates.

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `cargo +nightly check -p aws-sdk-ec2 --frozen` | 582.0 ± 4.3 | 576.3 |590.3 | 1.00 |
| `cargo +nightly check -p aws-sdk-ec2 --frozen -Zchecksum-freshness` | 698.1 ± 2.9 | 691.5 | 702.2 | 1.20 ± 0.01 |

<details><summary>Benchmark details</summary>
<p>

Setup package

```shell
git clone https://github.com/awslabs/aws-sdk-rust.git
cd aws-sdk-rust
git switch -d 505dab66bf0801ca743212678d47d6490d2beba9
```

Benchmark script

```shell
hyperfine --warmup 1 \
  --setup "cargo fetch && cargo clean" \
  --min-runs 20 \
  --export-markdown result.md \
  "cargo +nightly check -p aws-sdk-ec2 --frozen" \
  "cargo +nightly check -p aws-sdk-ec2 --frozen -Zchecksum-freshness
```

Platform information

```
cargo 1.82.0 (8f40fc59f 2024-08-21)
release: 1.82.0
commit-hash: 8f40fc59fb0c8df91c97405785197f3c630304ea
commit-date: 2024-08-21
host: x86_64-unknown-linux-gnu
libgit2: 1.8.1 (sys:0.19.0 vendored)
libcurl: 8.9.0-DEV (sys:0.4.74+curl-8.9.0 vendored ssl:OpenSSL/1.1.1w)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Amazon Linux AMI 2.0.0 [64-bit] (r7a.24xlarge)
```

</p>
</details> 

Additionally, here is the result of running the same benchmark but for the entire `aws-sdk-rust` workspace. `rg 'checksum:blake3' target/debug/deps | wc -l` shows 192323 matches.

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `cargo +nightly check --frozen` | 2.070 ± 0.099 | 2.035 | 2.492 | 1.00 |
| `cargo +nightly check --frozen -Zchecksum-freshness` | 5.063 ± 0.010 | 5.046 | 5.079 | 2.45 ± 0.12

> [!IMPORTANT]
> Benchmarkes were done on `r7a.24xlarge`, which has 96vCPU and 768GiB RAM. It is expected that on a lower end machine the performance will be worse.

**Comment by Xaeroxe** [comment]
Do the slower benchmarks appear to be I/O bound or CPU bound? If they are CPU bound, are all the cores saturated?

I knew going into this that performance would suffer. So unless we can improve these numbers a lot I'd be inclined to make this feature opt-in. Use it if you know you need it.

**Comment by epage** [comment]
> rg 'checksum:blake3' target/debug/deps | wc -l shows 8313 matches. It means rustc roughly computes blake3 checksum 8313 times.

Should we avoid passing this flag to rustc for immutable dependencies?  I know your benchmark didn't show much of a difference but figure there might be other relevant cases

**Comment by weihanglo** [comment]
Moved the performance discussion to #14722, and left this for tracking overall progress.

**Comment by epage** [comment]
I do not think we should block stabilization on build scripts. Unless we introduce TOCTOU race conditions, build scripts will always need to be able to support mtime freshness. That also means the transition for build scripts will be based on when the build script transitions, rather than based on any change in cargo, The most we can do is unblock it people transitioning. 

I lean towards doing the transition through an official config field. Because of build scripts, we'll still need to maintain the mtime code paths, so no internal advantage from rushing it. People may want to use mtimes for performance or benchmarking reasons.

If we do have a config, I feel like it would belong under `build`. The question is then naming. I feel like "freshness" isn't too clear.
- scons calls this `Decider`: https://scons.org/doc/production/HTML/scons-man.html#f-Decider
- shake calls this `shakeChange`: https://hackage-content.haskell.org/package/shake-0.19.9/docs/Development-Shake.html
- doit calls this `check_file_uptodate`: https://pydoit.org/cmd-run.html#check-file-uptodate
- ccache calls this `compiler_check`: https://ccache.dev/manual/latest.html#_configuration_options
- gradle calls this `up-to-date-check`, `track`

As for the values, I see
- `content`, `checksum`, `md5`, `digest`
- `timestamp`, `mtime`

build2 does things a little different: it checksums the tokens + their location, ignoring "whitespace" (which would include comments.  See https://build2.org/blog/faster-than-ninja.xhtml.  If we had relink, don't rebuild, the impact of whitespace only changes would be limited and those are unlikely enough that it might not be worth it.

build2 led to a crazy thought that might make things too slow: what if rustc had a flag for us to pass a bunch of files and it checksums however it wants, the same way it will checksum for depinfo, fully abstracting the "how" from callers. Calling rustc is expensive though and we might need to do this once per build unit unless we did some serious refactors.

**Comment by epage** [comment]
If we keep the config long term, we could forward it to build scripts so they can match the same style as cargo. Unsure if that would be worth it during a transition.

**Comment by epage** [comment]
While terms like "fingerprint", "track", and "trigger" are more general than just for files, I feel like the probability of a collision during the transition is low.

**Comment by epage** [comment]
To get the conversation going, I'm going to propose
```
[build[
fingerprint = "mtime". # also supports "content"
```

The steps would be
1. Implement the config
2. Stabilize
3. Resolve https://github.com/rust-lang/cargo/issues/14722
4. Wait a reasonable amount of time
5. Change the default to `"content"` or a hybrid mode developed for https://github.com/rust-lang/cargo/issues/14722

From there, we would evaluate any next steps, like deprecating and eventually removing the config field.

**Comment by ranger-ross** [comment]
Would also like to see this move forward. :)

bikeshed:

> fingerprint = "mtime". # also supports "content"

In a previous comment above, you mentioned some other alternatives like `timestamp`.
I kind of lean towards `timestamp` as it does not disclose the method of determining the timestamp to leaves us room to change the underlying implementation without the name becoming misleading.

**Comment by epage** [comment]
I don't think we can change to another timestamp source and still have this work.

**Comment by 0xdeafbeef** [comment]
Nix provides all files with a fixed mtime (1 second after the Unix epoch) for reproducibility. I've used `-Zchecksum-freshness`. It works great for regular source files, but build-script watches (`rerun-if-changed`, or the whole package by default) still check mtimes, which breaks persistent target caching: changed inputs may have the same mtime, so Cargo can miss them and reuse stale build-script output. I've added a hack where a small Python script preserves mtimes for unchanged files and advances them when their content changes, but maybe this could be part of Cargo's checksum freshness support?

I guess it's kinda niche request, but nevertheless :)

**Comment by bjorn3** [comment]
For compiled artifacts cargo could store the current time in a separate file next to the compiled artifacts right after rustc finishes. I don't see why we would do that, but it would probably be a working non-mtime timestamp source. For source files I do agree, mtime is the only timestamp option.

**Comment by epage** [comment]
> Nix provides all files with a fixed mtime (1 second after the Unix epoch) for reproducibility. I've used `-Zchecksum-freshness`. It works great for regular source files, but build-script watches (`rerun-if-changed`, or the whole package by default) still check mtimes, which breaks persistent target caching: changed inputs may have the same mtime, so Cargo can miss them and reuse stale build-script output. I've added a hack where a small Python script preserves mtimes for unchanged files and advances them when their content changes, but maybe this could be part of Cargo's checksum freshness support?
> 
> I guess it's kinda niche request, but nevertheless :)

Build scripts provide `re-run-if-changed=<path>` directives to cargo. This was designed assuming mtimes. We could calculate checksums for this but we would run into race conditions like
1. build script executes
2. file is changed
3. cargo reads checksum

On next build, cargo will see the checksum, see that nothing has changed, and use a stale execution of the build script. This might seem unlikely but think of cases like rust-analyzer running `cargo check` in the background while makkng edits.

For build scripts to work properly, we would need a new directive where the build script is able to tell Cargo the checksum of the content it read, removing the race condition.

Until the new directive is released and all build scripts have migrated to it, mtimes will still be in use if this feature gets stabilized as-is. To me, this means the features are decoupled enough that we shouldn't block progress on this feature for build script checksuming.

This is what I was referring to at https://github.com/rust-lang/cargo/issues/14136#issuecomment-5210212104

> I do not think we should block stabilization on build scripts. Unless we introduce TOCTOU race conditions, build scripts will always need to be able to support mtime freshness. That also means the transition for build scripts will be based on when the build script transitions, rather than based on any change in cargo, The most we can do is unblock it people transitioning.

**Comment by epage** [comment]
We talked about this in today's Cargo team meeting.

The only change from https://github.com/rust-lang/cargo/issues/14136#issuecomment-5210299257 is that we would block changing the default on better performance, whether by introducing a hybrid mode or gathering checksums in parallel, see https://github.com/rust-lang/cargo/issues/14722

@joshtriplett brought up some hybrid mtime heuristics seen in other systems and it would be good to bring that up on https://github.com/rust-lang/cargo/issues/14722 for consideration.  I'd recommend also pinging https://github.com/rust-lang/cargo/issues/12060 about it as the participants there may also have input on such a design.

**Comment by celinval** [comment]
For build script, can we extend the directive for build scripts to optionally include a checksum? Another option would be a hybrid approach where cargo uses the mtime as a way to identify if the checksum is fresh:
1. build script executes
2. cargo reads checksum
3. cargo reads last modified time and only record checksum if last modified time is before build script execution started.

**Comment by epage** [comment]
Improving build scripts is being decoupled from this because it needs a new directive that the ecosystem has to migrate to independent of this.

**Comment by celinval** [comment]
Is there an issue capturing the build script work?

**Comment by weihanglo** [comment]
There is no specific one for build script checksum freshenss. Feel free to create one.

**Comment by epage** [comment]
#17382 changed the shape of this feature from `-Zchecksum-freshness` enabling cheksums to respecting the new `build.fingerprint` field. It will take a few days to hit nightly.

To prepare, people should likely enable the new config field (which will otherwise be ignored) so they don't have to worry about when the specific nightly becomes available.

To enable checksum whenever you use a nightly toolchain, add the following to your `.cargo/config.toml` file:
```toml
[unstable]
checksum-freshness = true

[build]
fingerprint = "content"
```

**Comment by epage** [comment]
### Call for testing

We are preparing to stabilize a new `build.fingerprint` config field that would allow for overriding the default from `mtime` to `content`.

Caveats
- This does not apply to local build scripts, see [#t-cargo/build-script > Transitioning from mtimes to checksums @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/628857-t-cargo.2Fbuild-script/topic/Transitioning.20from.20mtimes.20to.20checksums/near/621211844)
- This can negatively impact build times which we are tracking in https://github.com/rust-lang/cargo/issues/14722

#### Basic configuration

Requirements:
- 1.100.0-nightly (2e2b193f8 2026-09-02)

Set in `.cargo/config.toml`:
```toml
[unstable]
checksum-freshness = true

[build]
fingerprint = "content"
```

#### Questions and areas of risk

- Does the naming make sense?
  - Prior art was collected in https://github.com/rust-lang/cargo/issues/14136#issuecomment-5210212104
- Use cases where `mtime` has functional benefits?
- Switching between `content` and `mtime` correctly rebuilds
- Correct handling of mixed fingerprinting methods between local source (`content`) and local build script inputs (mtime)

**Comment by U007D** [comment]
This RFC will appear in the **Call for Testing** section of the next issue (#668) of This Week in Rust (TWiR).
You may remove the `call-for-testing` label.  Please feel free to leave the `call-for-testing` label in place if you would like this RFC to appear again in another issue of TWiR.
